### PR TITLE
Refactor write segmentation I 

### PR DESCRIPTION
**Pull request by h-mayorquin**
This is part of a join effort with @weiglszonja to refactor `write_segmentation` in the roiextractors module. This first PR does a couple of things:
1) It refactors the previously added function called `add_imaging_plane` and adds some unit tests.
2) The `add_imaging_plane` function is then used to refactor the `write_segmentation` function with some code reduction.
3) There are some other minor refactoring in documentation (add_device function) and style here and there.

### CODE REVIEW

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py [hunk 1]
```diff
@@ -38,6 +39,20 @@ def get_default_ophys_metadata():
     metadata = get_default_nwbfile_metadata()
 
     default_device = dict(name="Microscope")
+    default_optical_channel = dict(
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:42]: This is just to flatten the metadata construction.

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py [hunk 2]
```diff
@@ -124,7 +123,12 @@ def get_nwb_imaging_metadata(imgextractor: ImagingExtractor):
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:123]: Flattening ends here.

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py [hunk 3]
```diff
@@ -138,27 +142,44 @@ def add_devices(nwbfile: NWBFile, metadata: dict) -> NWBFile:
...
+    imaging_plane_name = imaging_plane_metadata["name"]
+    existing_imaging_planes = nwbfile.imaging_planes
+
+    if imaging_plane_name not in existing_imaging_planes:
+        imaging_plane = _create_imaging_plane_from_metadata(
+            nwbfile=nwbfile, imaging_plane_metadata=imaging_plane_metadata
+        )
+        nwbfile.add_imaging_plane(imaging_plane)
 
     return nwbfile
```
Comment (h-mayorquin) [inline src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:178]: The diff is doing a terrible job at displaying this but from line 141 to here the code displays the substitution of the old private method `_add_imaging_plane` for a new `add_imaging_plane` which adds the imaging plane to an nwb file and a second private function `_create_imaging_plane_from_metadata` which creates an image plane from the metadata when its name does not exist in the nwbfile yet.

Comment (weiglszonja) [reply src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:178]: I like this approach :)

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py [hunk 4]
```diff
@@ -138,27 +142,44 @@ def add_devices(nwbfile: NWBFile, metadata: dict) -> NWBFile:
...
-    image_plane_metadata["device"] = nwbfile.devices[device_name]
+    imaging_plane = ImagingPlane(**imaging_plane_metadata)
 
-    image_plane_metadata["optical_channel"] = [
-        OpticalChannel(**metadata) for metadata in image_plane_metadata["optical_channel"]
-    ]
+    return imaging_plane
+
+
+def add_imaging_plane(nwbfile: NWBFile, metadata=dict, imaging_plane_index: int = 0) -> NWBFile:
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:159]: Yep.  This should be done, it was an oversight by me.

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py [hunk 5]
```diff
@@ -138,27 +142,44 @@ def add_devices(nwbfile: NWBFile, metadata: dict) -> NWBFile:
...
+
+    # Set the defaults and required infraestructure
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:161]: Thanks.

File: tests/test_internals/test_roi_extractors_module.py [hunk 6]
```diff
@@ -143,6 +144,92 @@ def test_device_object_and_metadata_mix(self):
...
+        self.imaging_plane_metadata["name"] = first_imaging_plane_name
+        self.imaging_plane_metadata["description"] = first_imaging_plane_description
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
+
+        # Add the second imaging plane
+        second_imaging_plane_name = "second_imaging_plane_name"
+        second_imaging_plane_description = "second_imaging_plane_description"
+        self.imaging_plane_metadata["name"] = second_imaging_plane_name
+        self.imaging_plane_metadata["description"] = second_imaging_plane_description
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply tests/test_internals/test_roi_extractors_module.py:218]: Yes, I think this is confusing and I think it stems from a design decision that was made about our metadata format. In order to mimic the nwb schema the metadata for imagine plane is stored at the top level of ophys as a list. The reason for this, as far as I understand, is that the `imaging_plane` in the nwb-schema is a group of objects.  Each of the imaging plane in this group are then stored as links in all the objects that require them (e.g. TwoPhotonSeries). I think that the decision was make to make our own schema in nwb-conversion tools mimic as much the true schema. 

The important outcome is that the metadata for the `imaging_plane` in our metadata dict is a list and therefore an specification of which element/index of the list do we want to add becomes necessary. That's the role of `imagine_plane_index`. If you see the previous code that this is removing it was assumed that the element was always the first which I think works right now because we never had a conversion with more than one plane. I changed this to make this assumption explicit and make it a parameter that can be changed if it is required. 

Maybe there is a better name? as we don't have examples of conversion data with more than two imaging planes I don't know what's the best way to handle this.

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/nwb_conversion_tools/tools/roiextractors/roiextractors.py]: I added them in the latest commit. Let me know what you think.

File: src/nwb_conversion_tools/tools/roiextractors/roiextractors.py (hunk 3 above)
Comment (h-mayorquin) [reply src/nwb_conversion_tools/tools/roiextractors/roiextractors.py:212]: Yes, I think that makes sense for this module. Having an auxiliary private function that creates the pynwb objects from the metadata and then a public tested function that adds this to the nwb file checking for edge cases and controlling parameters/options/etc.

**Comment by h-mayorquin** [comment]
@CodyCBakerPhD it seems that the windows tests are failing again even with re-running.

File: tests/test_internals/test_roi_extractors_module.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (weiglszonja) [reply tests/test_internals/test_roi_extractors_module.py:218]: I see, thanks for clarifying it.

**Review by weiglszonja — APPROVED** [review]

**Comment by CodyCBakerPhD** [comment]
@h-mayorquin @weiglszonja FYI this stuff did not make it over to https://github.com/catalystneuro/neuroconv prior to commit I used to split so this chain of PRs will have to be ported over
